caluma_client/models.py

```diff
@@ -0,0 +1,179 @@
+
+        self.document = document
+        self.raw = raw
+        self.form = None
+        self.fields = []
+
+        self._create_form()
+        self._create_fields()
+
+    def _create_form(self):
+        self.form = Form(self.raw.get("form"))
+
+    def _create_fields(self):
+        def _find_answer(question):
+            return next(
+                (
+                    answer
+                    for answer in self.raw["answers"]
+                    if answer.get("question", {}).get("slug") == question.get("slug")
+                ),
+                None,
+            )
+
+        self.fields = [
+            Field(self, {"question": question, "answer": _find_answer(question)})
+            for question in self.raw["form"]["questions"]
+        ]
+
+    @property
+    def field(self):
+        raise NotImplementedError
+
+
+class Field:
+    def __init__(self, fieldset, raw):
+        assert "question" in raw, "Raw must contain Question"
+
+        self.fieldset = fieldset
+        self.raw = raw
+
+        self.question = Question(self.raw["question"])
+        self.answer = None
+        self.options = None  # TODO?
+        self._create_answer()
+
+    def _create_answer(self):
+        if self.raw.get("answer"):
+            self.answer = Answer(self.raw["answer"])
+            return
+
+        question_type = self.raw["question"]["__typename"]
+        answer_type = ANSWER_TYPE_MAP.get(question_type)
+        if answer_type is None:
+            return
+
+        value_key = answer_value_key(question_type)
+        raw = dict(
+            zip(
+                ("__typename", "question", value_key),
+                (answer_type, self.raw["question"].get("slug"), None),
+            )
+        )
+        self.answer = Answer(raw)
+
+
+class Form:
+    def __init__(self, raw):
+        assert raw.get("__typename") == "Form", "raw must be a caluma `Form`"
+
+        self.raw = raw
+
+    @property
+    def uuid(self):
+        return decode_id(self.raw["id"]) if "id" in self.raw else None
+
+
+class Question:
+    def __init__(self, raw):
+        assert raw.get("__typename").endswith(
+            "Question"
+        ), "raw must be a caluma `Question`"
+        self.raw = raw
+        # self.is_choice = raw["__typename"].
+
+
+class Answer:
+    def __init__(self, raw):
+        assert raw.get("__typename", "").endswith(
+            "Answer"
+        ), "raw must be a caluma `Answer`"
+        self.raw = raw
+        self.__typename = raw["__typename"]
+        self.value_key = answer_value_key(self.raw["__typename"])
+        self._create_value()
+
+    def _create_value(self):
+        value = self.raw.get(self.value_key)
+
+        if self.__typename == "TableAnswer" and value:
+            self.value = [Document(parse_document(raw)) for raw in value]
+        else:
+            self.value = value
+
+    @property
+    def uuid(self):
+        return decode_id(self.raw["id"]) if "id" in self.raw else None
```

Comment:
I like that approach. Maybe something like this would make creating @properties faster:

```python
def __init__(self....):
    ...
    self.uuid = create_getter(type='uuid', dom_path='id', default=None)
    ...

create_getter returns a closure that returns the right data from raw.

Reply:
The `uuid` and other fields with `@property` are more or less a left-over from porting the `ember-caluma/addon/lib` code. There, a lot of fields need to be dynamically reloadable when there are updates. But since this library doesn't do updates, we could purge the `@property` usages altogether.

Reply:
I my idea was more about generating path accessors and conversions, so the definition of properties / fields or - what ever you want to call it - is stictly declarative and the logic is completely hidden in create_getter. Then it feels like a schema definition. It all depends how many different types of fields and how complicated the dom-path can become. It might be total overkill or not... dunno.

Reply:
That was addressed in the recent commit